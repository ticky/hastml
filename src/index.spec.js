/* global jest, describe, expect, it */
import walker from '.';

describe('walker', () => {
  it('throws an error if no callback is supplied', () => {
    expect(() => walker('<html />')).toThrowErrorMatchingSnapshot();
  });

  it('calls the callback for each given HTML token', () => {
    const callback = jest.fn(
      (...args) => {
        expect(args).toMatchSnapshot();
        return args[0];
      }
    );

    const output = walker(
      '<html><head></head><body><img /><!-- ignore tis plx--></body></html>',
      callback
    );

    expect(callback.mock.calls.length).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

  it(`doesn't mutate the output if the callback doesn't explicitly return a string value`, () => {
    const callback = jest.fn(
      (...args) => {
        expect(args).toMatchSnapshot();
      }
    );

    const output = walker(
      '<html><head></head><body><img /><!-- ignore tis plx--></body></html>',
      callback
    );

    expect(callback.mock.calls.length).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

  it('gives correct coordinates when the HTML is changed', () => {
    const callback = jest.fn(
      (...args) => {
        expect(args).toMatchSnapshot();
        return `${args[0]} autocomplete="off"`;
      }
    );

    const output = walker(
      '<html><head></head><body><img /><!-- ignore tis plx--></body></html>',
      callback
    );

    expect(callback.mock.calls.length).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });
});
