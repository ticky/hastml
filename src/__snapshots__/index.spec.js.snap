// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walker calls the callback for each given HTML token 1`] = `
Array [
  "<html",
  "<",
  0,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 0,
    "state": "open",
    "tagName": "html",
  },
  Array [
    Object {
      "openIndex": 0,
      "state": "open",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 2`] = `
Array [
  ">",
  ">",
  5,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 5,
    "openIndex": 0,
    "state": "content",
    "tagName": "html",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 3`] = `
Array [
  "<head",
  "<",
  6,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 6,
    "state": "open",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "openIndex": 6,
      "state": "open",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 4`] = `
Array [
  ">",
  ">",
  11,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 11,
    "openIndex": 6,
    "state": "content",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 11,
      "openIndex": 6,
      "state": "content",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 5`] = `
Array [
  "</head",
  "</",
  12,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 12,
    "contentIndex": 11,
    "openIndex": 6,
    "state": "closing",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closingIndex": 12,
      "contentIndex": 11,
      "openIndex": 6,
      "state": "closing",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 6`] = `
Array [
  ">",
  ">",
  18,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 18,
    "closingIndex": 12,
    "contentIndex": 11,
    "openIndex": 6,
    "state": "closing",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closeIndex": 18,
      "closingIndex": 12,
      "contentIndex": 11,
      "openIndex": 6,
      "state": "closing",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 7`] = `
Array [
  "<body",
  "<",
  19,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 19,
    "state": "open",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "openIndex": 19,
      "state": "open",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 8`] = `
Array [
  ">",
  ">",
  24,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 24,
    "openIndex": 19,
    "state": "content",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 9`] = `
Array [
  "<img",
  "<",
  25,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 25,
    "state": "open",
    "tagName": "img",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
    Object {
      "openIndex": 25,
      "state": "open",
      "tagName": "img",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 10`] = `
Array [
  "/>",
  "/>",
  30,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 30,
    "openIndex": 25,
    "state": "open",
    "tagName": "img",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
    Object {
      "closeIndex": 30,
      "openIndex": 25,
      "state": "open",
      "tagName": "img",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 11`] = `
Array [
  "</body",
  "</",
  32,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 32,
    "contentIndex": 24,
    "openIndex": 19,
    "state": "closing",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closingIndex": 32,
      "contentIndex": 24,
      "openIndex": 19,
      "state": "closing",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 12`] = `
Array [
  ">",
  ">",
  38,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 38,
    "closingIndex": 32,
    "contentIndex": 24,
    "openIndex": 19,
    "state": "closing",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closeIndex": 38,
      "closingIndex": 32,
      "contentIndex": 24,
      "openIndex": 19,
      "state": "closing",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 13`] = `
Array [
  "</html",
  "</",
  39,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 39,
    "contentIndex": 5,
    "openIndex": 0,
    "state": "closing",
    "tagName": "html",
  },
  Array [
    Object {
      "closingIndex": 39,
      "contentIndex": 5,
      "openIndex": 0,
      "state": "closing",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 14`] = `
Array [
  ">",
  ">",
  45,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 45,
    "closingIndex": 39,
    "contentIndex": 5,
    "openIndex": 0,
    "state": "closing",
    "tagName": "html",
  },
  Array [
    Object {
      "closeIndex": 45,
      "closingIndex": 39,
      "contentIndex": 5,
      "openIndex": 0,
      "state": "closing",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker calls the callback for each given HTML token 15`] = `14`;

exports[`walker calls the callback for each given HTML token 16`] = `"<html><head></head><body><img /></body></html>"`;

exports[`walker gives correct coordinates when the HTML is changed 1`] = `
Array [
  "<html",
  "<",
  0,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 0,
    "state": "open",
    "tagName": "html",
  },
  Array [
    Object {
      "openIndex": 0,
      "state": "open",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 2`] = `
Array [
  ">",
  ">",
  5,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 5,
    "openIndex": 0,
    "state": "content",
    "tagName": "html",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 3`] = `
Array [
  "<head",
  "<",
  6,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 6,
    "state": "open",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "openIndex": 6,
      "state": "open",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 4`] = `
Array [
  ">",
  ">",
  11,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 11,
    "openIndex": 6,
    "state": "content",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 11,
      "openIndex": 6,
      "state": "content",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 5`] = `
Array [
  "</head",
  "</",
  12,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 12,
    "contentIndex": 11,
    "openIndex": 6,
    "state": "closing",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closingIndex": 12,
      "contentIndex": 11,
      "openIndex": 6,
      "state": "closing",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 6`] = `
Array [
  ">",
  ">",
  18,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 18,
    "closingIndex": 12,
    "contentIndex": 11,
    "openIndex": 6,
    "state": "closing",
    "tagName": "head",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closeIndex": 18,
      "closingIndex": 12,
      "contentIndex": 11,
      "openIndex": 6,
      "state": "closing",
      "tagName": "head",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 7`] = `
Array [
  "<body",
  "<",
  19,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 19,
    "state": "open",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "openIndex": 19,
      "state": "open",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 8`] = `
Array [
  ">",
  ">",
  24,
  "<html><head></head><body><img /></body></html>",
  Object {
    "contentIndex": 24,
    "openIndex": 19,
    "state": "content",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 9`] = `
Array [
  "<img",
  "<",
  25,
  "<html><head></head><body><img /></body></html>",
  Object {
    "openIndex": 25,
    "state": "open",
    "tagName": "img",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
    Object {
      "openIndex": 25,
      "state": "open",
      "tagName": "img",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 10`] = `
Array [
  "/>",
  "/>",
  30,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 30,
    "openIndex": 25,
    "state": "open",
    "tagName": "img",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "contentIndex": 24,
      "openIndex": 19,
      "state": "content",
      "tagName": "body",
    },
    Object {
      "closeIndex": 30,
      "openIndex": 25,
      "state": "open",
      "tagName": "img",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 11`] = `
Array [
  "</body",
  "</",
  32,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 32,
    "contentIndex": 24,
    "openIndex": 19,
    "state": "closing",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closingIndex": 32,
      "contentIndex": 24,
      "openIndex": 19,
      "state": "closing",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 12`] = `
Array [
  ">",
  ">",
  38,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 38,
    "closingIndex": 32,
    "contentIndex": 24,
    "openIndex": 19,
    "state": "closing",
    "tagName": "body",
  },
  Array [
    Object {
      "contentIndex": 5,
      "openIndex": 0,
      "state": "content",
      "tagName": "html",
    },
    Object {
      "closeIndex": 38,
      "closingIndex": 32,
      "contentIndex": 24,
      "openIndex": 19,
      "state": "closing",
      "tagName": "body",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 13`] = `
Array [
  "</html",
  "</",
  39,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closingIndex": 39,
    "contentIndex": 5,
    "openIndex": 0,
    "state": "closing",
    "tagName": "html",
  },
  Array [
    Object {
      "closingIndex": 39,
      "contentIndex": 5,
      "openIndex": 0,
      "state": "closing",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 14`] = `
Array [
  ">",
  ">",
  45,
  "<html><head></head><body><img /></body></html>",
  Object {
    "closeIndex": 45,
    "closingIndex": 39,
    "contentIndex": 5,
    "openIndex": 0,
    "state": "closing",
    "tagName": "html",
  },
  Array [
    Object {
      "closeIndex": 45,
      "closingIndex": 39,
      "contentIndex": 5,
      "openIndex": 0,
      "state": "closing",
      "tagName": "html",
    },
  ],
]
`;

exports[`walker gives correct coordinates when the HTML is changed 15`] = `14`;

exports[`walker gives correct coordinates when the HTML is changed 16`] = `"<html autocomplete=\\"off\\"> autocomplete=\\"off\\"<head autocomplete=\\"off\\"> autocomplete=\\"off\\"</head autocomplete=\\"off\\"> autocomplete=\\"off\\"<body autocomplete=\\"off\\"> autocomplete=\\"off\\"<img autocomplete=\\"off\\" /> autocomplete=\\"off\\"</body autocomplete=\\"off\\"> autocomplete=\\"off\\"</html autocomplete=\\"off\\"> autocomplete=\\"off\\""`;

exports[`walker throws an error if no callback is supplied 1`] = `"A callback is required!"`;
